### PR TITLE
ServiceNow Incident SNS Topic

### PR DESCRIPTION
**Pull request by pauldoomgov**
SNS topic and Lambda to POST an Incident to ServiceNow upon an event.

From README:
# `snow_incident` - ServiceNow Incident Generation SNS Topic

TL;DR - Create a ServiceNow Incident when a CloudWatch alarm (or other event) is triggered.

This Terraform module is designed to create a Lambda function, containing Python code, which will POST a message to ServiceNow and raise an incident.

Created resources:
- Lambda function + IAM role/policy
- Empty SSM snow_username and snow_password parameters (See below how to set values)
- CloudWatch log group
- SNS execution permission for the Lambda
- SNS topic
- SNS topic subscription

### Example Module

~~~hcl
module "snow_incident_topic" {
  source = "github.com/18F/identity-terraform/snow_incident?ref=main"

  name                = "snow_incident_topic"
  topic_name          = "snow-incident"
  snow_incident_url   = "https://something.servicenowservices.com/api/now/table/incident?sysparm_fields=number%2Cincident_state%2Cshort_description%2Cu_csi%2Cassignment_group"
  snow_caller_id      = "deadbeef000011112222"
  snow_category_id    = "1234cafe4321"
  snow_subcategory_id = "1234567890987654321"
  snow_item_id        = "0987654321234567890"
  snow_parameter_base = "/account/snow_incident"
}
~~~

### Terraform Variables

Unfortunately a lot of parameters must be set to use this module.   Fortunately most should be provided
by ServiceNow as part of integration.

* `name` - Name of the Lambda function.
* `topic_name` - SNS topic name
* `snow_incident_url` - ServiceNow URL to POST to (maps to `SNOW_INCIDENT_URL` environment variable)
* `snow_caller_id` - ServiceNow ID for Caller ID entity (maps to `SNOW_CALLER_ID` environment variable)
* `snow_category_id` - ServiceNow ID for category (maps to `SNOW_CATEGORY_ID` environment variable)
* `snow_subcategory_id` - ServiceNow ID for sub-category (maps to `SNOW_SUBCATEGORY_ID` environment variable)
* `snow_item_id` - ServiceNow ID for item (maps to `SNOW_ITEM_ID` environment variable)
* `snow_parameter_base` - SSM Parameter Store base path, starting with `/` - The `snow_username` and `snow_password` should be set under this path (maps to `SNOW_PARAMETER_BASE` environment variable)

### CODE REVIEW

File: snow_incident/snow_incident.py [hunk 1]
```diff
@@ -0,0 +1,199 @@
...
+        full_error = (
+            f"SNOW request failed with status {response.status_code}, ",
+            f"Headers: {response.headers}, ",
+            f"Response: {response.data.decode('utf-8')}",
+        )
+        raise ValueError(full_error)
+
+    return json.loads(response.data.decode("utf-8"))
+
+    return data["number"]
```
Comment (akrito) [inline snow_incident/snow_incident.py:175]: `data` isn't defined in this context, and also this `return` statement won't be called.

**Review by akrito — CHANGES_REQUESTED** [review]

**Comment by ryan-brown-sb** [comment]
branch was merged

**Comment by pauldoomgov** [comment]
We did not move to using ServiceNow for alerting.
